### PR TITLE
Custom query fix

### PR DESCRIPTION
**Pull request by mhavey**
*Issue #, if available:*

https://github.com/awslabs/aws-athena-query-federation/issues/1398

*Description of changes:*
Add RDF support to Neptune connector. This builds on recent changes for property graph views in Neptune connector.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: athena-neptune/src/main/java/com/amazonaws/athena/connectors/neptune/NeptuneRecordHandler.java [hunk 1]
```diff
@@ -59,18 +61,37 @@ public class NeptuneRecordHandler extends RecordHandler
...
      */
     private static final String SOURCE_TYPE = "neptune";
-    private final NeptuneConnection neptuneConnection;
+    private NeptuneConnection neptuneConnection = null;
+    private java.util.Map<String, String> configOptions = null;
+
+    private static NeptuneConnection createConnection(java.util.Map<String, String> configOptions) 
+    {
+        GraphType graphType = GraphType.PROPERTYGRAPH;
+        if (System.getenv("neptune_graphtype") != null) {
```
Comment (burhan94) [inline athena-neptune/src/main/java/com/amazonaws/athena/connectors/neptune/NeptuneRecordHandler.java:70]: use configOptions.getEnv here. We dump System.getEnv to configOptions when the RecordHandler is instantiated.

File: athena-neptune/src/main/java/com/amazonaws/athena/connectors/neptune/NeptuneRecordHandler.java [hunk 2]
```diff
@@ -59,18 +61,37 @@ public class NeptuneRecordHandler extends RecordHandler
...
+            graphType = GraphType.valueOf(System.getenv("neptune_graphtype").toUpperCase());
+        }
+
+        switch(graphType){
+            case PROPERTYGRAPH: 
+                return new NeptuneGremlinConnection(configOptions.get("neptune_endpoint"),
+                configOptions.get("neptune_port"), Boolean.parseBoolean(configOptions.get("iam_enabled")),
```
Comment (burhan94) [inline athena-neptune/src/main/java/com/amazonaws/athena/connectors/neptune/NeptuneRecordHandler.java:77]: lets make static variables for these properties.

File: athena-neptune/src/main/java/com/amazonaws/athena/connectors/neptune/propertygraph/PropertyGraphHandler.java [hunk 3]
```diff
@@ -135,23 +139,44 @@ public void executeQuery(
                         EdgeRowWriter.writeRowTemplate(builder, nextField, configOptions);
                     }
 
+                    parseNodeOrEdge(queryStatusChecker, spiller, numRows, graphTraversal, builder);
+
+                    break;
+                    
+                case VIEW: 
+                    String query = recordsRequest.getSchema().getCustomMetadata().get("query");
```
Comment (burhan94) [inline athena-neptune/src/main/java/com/amazonaws/athena/connectors/neptune/propertygraph/PropertyGraphHandler.java:147]: add static variable here pls. Maybe its worth creating a class for constants.

File: athena-neptune/src/main/java/com/amazonaws/athena/connectors/neptune/rdf/NeptuneSparqlConnection.java [hunk 4]
```diff
@@ -0,0 +1,211 @@
...
+    RepositoryConnection connection = null;
+    TupleQueryResult queryResult = null;
+
+    public NeptuneSparqlConnection(String neptuneEndpoint, String neptunePort, boolean enabledIAM, String region) 
+    {
+        super(neptuneEndpoint, neptunePort, enabledIAM, region);
+        this.connectionString = "https://" + neptuneEndpoint + ":" + neptunePort;
+    }
+
+    public void connect() throws NeptuneSigV4SignerException
```
Comment (burhan94) [inline athena-neptune/src/main/java/com/amazonaws/athena/connectors/neptune/rdf/NeptuneSparqlConnection.java:62]: why does this need a separate connect() class? why not include this as a part of the constructor?

**Review by burhan94 — CHANGES_REQUESTED** [review]

File: athena-neptune/src/main/resources/log4j2.xml [hunk 5]
```diff
@@ -29,7 +29,8 @@
     </Appenders>
     <Loggers>
         <Logger name="com.amazonaws.athena.connector.lambda" level="${env:ATHENA_FEDERATION_SDK_LOG_LEVEL:-warn}" />
-        <Logger name="com.amazonaws.athena.connectors.neptune" level="${env:ATHENA_FEDERATION_SDK_LOG_LEVEL:-warn}" />
+        <Logger name="com.amazonaws.athena.connectors.neptune" level="warn" />
```
Comment (burhan94) [inline athena-neptune/src/main/resources/log4j2.xml:32]: please revert the changes and keep as: `level="${env:ATHENA_FEDERATION_SDK_LOG_LEVEL:-warn}`

This is used for customer debugging so that log levels can be configured via lambda env variable

File: athena-neptune/src/main/java/com/amazonaws/athena/connectors/neptune/rdf/RDFHandler.java [hunk 6]
```diff
@@ -0,0 +1,180 @@
...
+    {
+        // 1. Get the specified prefixes
+        Map<String, String> prefixMap = new HashMap<>();
+        String prefixBlock = "";
+        for (String k : recordsRequest.getSchema().getCustomMetadata().keySet()) {
+            if (k.startsWith(PREFIX_KEY)) {
+                String pfx = k.substring(PREFIX_LEN);
+                String val = recordsRequest.getSchema().getCustomMetadata().get(k);
+                prefixMap.put(pfx, val);
+                prefixBlock += "PREFIX " + pfx + ": <" + val + ">\n";
```
Comment (burhan94) [inline athena-neptune/src/main/java/com/amazonaws/athena/connectors/neptune/rdf/RDFHandler.java:111]: Please use a stringbuilder for readability or string.format.

File: athena-neptune/src/main/java/com/amazonaws/athena/connectors/neptune/propertygraph/rowwriters/CustomSchemaRowWriter.java [hunk 7]
```diff
@@ -0,0 +1,257 @@
...
+    private CustomSchemaRowWriter() 
+    {
+        // Empty private constructor
+    }
+
+    public static void writeRowTemplate(RowWriterBuilder rowWriterBuilder, Field field) 
+    {
+        ArrowType arrowType = field.getType();
+        Types.MinorType minorType = Types.getMinorTypeForArrowType(arrowType);
+        Boolean enableCaseinsensitivematch = (System.getenv("enable_caseinsensitivematch") == null) ? true : Boolean.parseBoolean(System.getenv("enable_caseinsensitivematch"));
```
Comment (burhan94) [inline athena-neptune/src/main/java/com/amazonaws/athena/connectors/neptune/propertygraph/rowwriters/CustomSchemaRowWriter.java:63]: you'll need to use configOptions here. Using System.getEnv within the connectors isn't recommended.

File: athena-neptune/src/main/java/com/amazonaws/athena/connectors/neptune/rdf/RDFHandler.java [hunk 8]
```diff
@@ -0,0 +1,180 @@
...
+            }
+        }
+
+        // 2. Build the SPARQL query
+        String queryMode = recordsRequest.getSchema().getCustomMetadata().get("querymode");
```
Comment (burhan94) [inline athena-neptune/src/main/java/com/amazonaws/athena/connectors/neptune/rdf/RDFHandler.java:116]: pls use constants for all of these. Its hard to determine where they're coming from without prior context.

**Review by burhan94 — CHANGES_REQUESTED** [review]

File: athena-neptune/src/main/java/com/amazonaws/athena/connectors/neptune/NeptuneRecordHandler.java (hunk 1 above)
Comment (mhavey) [reply athena-neptune/src/main/java/com/amazonaws/athena/connectors/neptune/NeptuneRecordHandler.java]: changed

File: athena-neptune/src/main/java/com/amazonaws/athena/connectors/neptune/NeptuneRecordHandler.java (hunk 2 above)
Comment (mhavey) [reply athena-neptune/src/main/java/com/amazonaws/athena/connectors/neptune/NeptuneRecordHandler.java]: changed

File: athena-neptune/src/main/java/com/amazonaws/athena/connectors/neptune/rdf/NeptuneSparqlConnection.java (hunk 4 above)
Comment (mhavey) [reply athena-neptune/src/main/java/com/amazonaws/athena/connectors/neptune/rdf/NeptuneSparqlConnection.java]: connect now from constructor

File: athena-neptune/src/main/resources/log4j2.xml (hunk 5 above)
Comment (mhavey) [reply athena-neptune/src/main/resources/log4j2.xml]: changed

File: athena-neptune/src/main/java/com/amazonaws/athena/connectors/neptune/rdf/RDFHandler.java (hunk 6 above)
Comment (mhavey) [reply athena-neptune/src/main/java/com/amazonaws/athena/connectors/neptune/rdf/RDFHandler.java]: now using string builder

File: athena-neptune/src/main/java/com/amazonaws/athena/connectors/neptune/propertygraph/rowwriters/CustomSchemaRowWriter.java (hunk 7 above)
Comment (mhavey) [reply athena-neptune/src/main/java/com/amazonaws/athena/connectors/neptune/propertygraph/rowwriters/CustomSchemaRowWriter.java]: changed

File: athena-neptune/src/main/java/com/amazonaws/athena/connectors/neptune/rdf/RDFHandler.java (hunk 8 above)
Comment (mhavey) [reply athena-neptune/src/main/java/com/amazonaws/athena/connectors/neptune/rdf/RDFHandler.java]: changed

**Comment by mhavey** [comment]
Made code changes to address review comments

**Comment by chngpe** [comment]
build failure 
`Error:  Failed to execute goal org.apache.maven.plugins:maven-compiler-plugin:3.11.0:compile (default-compile) on project athena-neptune: Compilation failure
Error:  /home/runner/work/aws-athena-query-federation/aws-athena-query-federation/athena-neptune/src/main/java/com/amazonaws/athena/connectors/neptune/propertygraph/PropertyGraphHandler.java:[30,67] package com.amazonaws.athena.connectors.neptune.propertygraph.Enums does not exist`

**Review by burhan94 — CHANGES_REQUESTED** [review]

File: athena-neptune/docs/aws-glue-sample-scripts/README.md [hunk 9]
```diff
@@ -40,46 +57,14 @@ Here's a reference of the Glue DataTypes that you can use:
...
-```
+For more on creating tables for property graph data, see [PropertyGraph.md](PropertyGraph.md).
 
-### Create AWS Glue Database and Tables manually
+For more on creating tables for RDF data, see [RDF.md](RDF.md).
 
 
-If you want to create database and tables manually, you can use the sample shell script [here](./manual/sample-cli-script.sh) to create a Glue Database by the name "graph-database" and tables: airport, country, continent and route  corresponding to the Air Routes Property Graph sample dataset. 
 
+<<<<<<< HEAD
```
[the remark this answers is not in the document]
Comment (mhavey) [reply athena-neptune/docs/aws-glue-sample-scripts/README.md]: Removed merge headers

**Review by burhan94 — APPROVED** [review]
